File: output/GPT4o-Mini/xCodeEval_4000/C++/Python/f5eedfa3d3ba232ace3697125d6ca9fc.py
import sys
from functools import lru_cache

sys.setrecursionlimit(200000)

USUAL_MOD = 1000000007
fftmod = 998244353

def powmod(x, n, m):
    res = 1
    x %= m
    while n:
        if n & 1:
            res = res * x % m
        x = x * x % m
        n //= 2
    return res

def gcd(a, b):
    while b:
        a, b = b, a % b
    return a

def read_int():
    return int(sys.stdin.readline().strip())

def read_ints():
    return list(map(int, sys.stdin.readline().strip().split()))

def cnk(n, k, fact, revFact):
    return fact[n] * revFact[k] % fftmod * revFact[n - k] % fftmod

def solve():
    global a
    fact = [1] * (200001)
    revFact = [1] * (202021)

    for i in range(1, 200001):
        fact[i] = fact[i - 1] * i % fftmod

    revFact[200000] = powmod(fact[200000], fftmod - 2, fftmod)
    for i in range(199999, -1, -1):
        revFact[i] = revFact[i + 1] * (i + 1) % fftmod

    n, k = read_ints()
    a = read_ints() + [0]  # Append a[0] at the end

    cnt = sum(a[i] != a[i + 1] for i in range(n))

    if k == 1 or n == 1:
        print(0)
        return

    ans = 0
    freeChoices = k - 2
    for i in range(cnt // 2 + 1):
        cur = (cnk(cnt, i, fact, revFact) * cnk(cnt - i, i, fact, revFact) % fftmod *
                powmod(freeChoices, cnt - 2 * i, fftmod) % fftmod *
                powmod(k, n - cnt, fftmod) % fftmod)
        ans += cur

    ans %= fftmod
    tot = powmod(k, n, fftmod) - ans
    if tot < 0:
        tot += fftmod

    print(tot * powmod(2, fftmod - 2, fftmod) % fftmod)

if __name__ == "__main__":
    solve()

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Java/f5eedfa3d3ba232ace3697125d6ca9fc.java
import java.io.*;
import java.util.*;

public class f5eedfa3d3ba232ace3697125d6ca9fc_nc {
    static final int fftmod = 998244353;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter pw = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));

        int[] a = new int[1010101];
        int[] fact = new int[201010];
        int[] revFact = new int[202020];

        fact[0] = 1;
        int N = 200000;
        for (int i = 1; i <= N; ++i) {
            fact[i] = (int) ((long) fact[i - 1] * i % fftmod);
        }
        revFact[N] = powmod(fact[N], fftmod - 2, fftmod);
        for (int i = N - 1; i >= 0; --i) {
            revFact[i] = (int) ((long) revFact[i + 1] * (i + 1) % fftmod);
        }

        String[] input = br.readLine().split(" ");
        int n = Integer.parseInt(input[0]);
        int k = Integer.parseInt(input[1]);

        input = br.readLine().split(" ");
        for (int i = 0; i < n; ++i) {
            a[i] = Integer.parseInt(input[i]);
        }
        a[n] = a[0];
        int cnt = 0;
        for (int i = 0; i < n; ++i) {
            if (a[i] != a[i + 1]) {
                cnt++;
            }
        }
        if (k == 1 || n == 1) {
            pw.println(0);
            pw.close();
            return;
        }
        long ans = 0;
        int freeChoices = k - 2;
        for (int i = 0; i <= cnt / 2; ++i) {
            long cur = (long) cnk(cnt, i) * cnk(cnt - i, i) % fftmod * powmod(freeChoices, cnt - 2 * i, fftmod) % fftmod * powmod(k, n - cnt, fftmod) % fftmod;
            ans += cur;
        }
        ans %= fftmod;
        long tot = powmod(k, n, fftmod) - ans;
        if (tot < 0) tot += fftmod;
        pw.println(tot * powmod(2, fftmod - 2, fftmod) % fftmod);
        pw.close();
    }

    static long powmod(long x, long n, long m) {
        long res = 1;
        x %= m;
        while (n > 0) {
            if ((n & 1) == 1) {
                res = res * x % m;
            }
            x = x * x % m;
            n >>= 1;
        }
        return res;
    }

    static long cnk(int n, int k) {
        long res = 1;
        for (int i = 0; i < k; ++i) {
            res = res * (n - i) % fftmod * powmod(i + 1, fftmod - 2, fftmod) % fftmod;
        }
        return res;
    }
}
